File: naddu/naddu_2/src/perft.js
// uses the move iterator and tt to exercise them.
// it also compares the incremental hash to a rebuilt hash

const TT_PERFT = 4;

function perft (depth, ply) {

  if (ply === 0)
    ttClear();

  if (depth === 0)
    return 1;

  const node = nodes[ply];
  const nextNode = nodes[ply+1];
  const pos = node.pos;
  const nextPos = nextNode.pos;
  const stmi = pos.stm >> 3;

  // check incremental hash matches rebuilt hash
  const hashLo = pos.hashLo;
  const hashHi = pos.hashHi;
  zobRebuild(pos)
  if (hashLo != pos.hashLo)
    console.log('*********** lo', hashLo, pos.hashLo);
  if (hashHi != pos.hashHi)
    console.log('************hi', hashHi, pos.hashHi);

  // probe tt
  const ttIndex = ttGet(pos);
  if (ttIndex >= 0 && ttGetType(ttIndex) === TT_PERFT && ttGetDepth(ttIndex) === depth) {
    return ttGetMove(ttIndex); // node count stored in move field
  }

  let tot = 0;

  initNextMove(node, 0);

  let move;

  while ((move = getNextMove(node))) {
    posSet(nextPos, pos);
    makeMove(move, nextPos);
    if (isAttacked(nextPos, nextPos.kings[stmi], nextPos.stm))
      continue;
    tot += perft(depth-1, ply+1);
  }

  if (depth > 2) // false positives otherwise
    ttPut(pos, TT_PERFT, depth, 0, tot);

  return tot;

}

File: naddu/naddu_2/c/perft.c
#include <stdio.h>
#include "perft.h"
#include "types.h"
#include "node.h"
#include "pos.h"
#include "tt.h"
#include "zob.h"
#include "iterate.h"
#include "make.h"
#include "attack.h"

#define TT_PERFT 4

uint64_t perft(int depth, int ply) {
  if (ply == 0)
    tt_clear();

  if (depth == 0)
    return 1;

  Node *node = &nodes[ply];
  Node *next_node = &nodes[ply + 1];
  Pos *pos = &node->pos;
  Pos *next_pos = &next_node->pos;
  int stmi = pos->stm >> 3;

  // Check incremental hash matches rebuilt hash
  uint64_t hash = pos->hash;
  zob_rebuild(pos);
  if (hash != pos->hash) {
    printf("*********** hash mismatch %016llx %016llx\n",
           (unsigned long long)hash, (unsigned long long)pos->hash);
  }

  // Probe TT
  int tt_index = tt_get(pos);
  if (tt_index >= 0 && tt_get_type(tt_index) == TT_PERFT && tt_get_depth(tt_index) == depth) {
    // Node count stored in move field (and score for high bits if needed)
    return tt_get_move(tt_index);
  }

  uint64_t tot = 0;

  init_next_move(node, 0);

  uint32_t move;
  while ((move = get_next_move(node))) {
    pos_copy(next_pos, pos);
    make_move(move, next_pos);

    if (is_attacked(next_pos, next_pos->kings[stmi], next_pos->stm))
      continue;

    tot += perft(depth - 1, ply + 1);
  }

  if (depth > 2)
    tt_put(pos, TT_PERFT, depth, 0, (uint32_t)tot);

  return tot;
}
